src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java

```diff
@@ -0,0 +1,160 @@
+	 *
+	 * @param clazz The NBTSerializable class to unregister.
+	 *
+	 * @throws IllegalArgumentException Throws if the given class to unregister is null.
+	 * */
+    public static <T extends NBTSerializable> void unregisterNBTSerializable(Class<T> clazz) {
+    	if (clazz == null) {
+			throw new IllegalArgumentException("Cannot unregister NBTSerializable: " +
+					"the given class is null.");
+		}
+		System.out.println("NBTSerializable[" + clazz.getName() + "] unregistered.");
+        registeredClasses.values().remove(clazz);
+    }
+
+    /**
+	 * Unregisters all {@link NBTSerializable} classes.
+	 * */
+    public static void clearAllRegistrations() {
+    	System.out.println("All NBTSerialization registered classes have been cleared.");
+    	registeredClasses.clear();
+	}
+
+	/**
+	 * Serializes an {@link NBTSerializable} class into an {@link NBTCompound} for the purpose of being able to
+	 * deserialize it later. For this purpose, {@link NBTSerializable#serialize(NBTCompound) NBTSerializable's
+	 * serialise} method should not be called directly.
+	 *
+	 * @param serializable The {@link NBTSerializable} class to serialize.
+	 * @return Returns an {@link NBTCompound} with the serialised data, or null if the given object was null, or if an
+	 * error occurred during serialization.
+	 * */
+    public static <T extends NBTSerializable> NBTCompound serialize(T serializable) {
+        if (serializable == null) {
+            return null;
+        }
+        NBTCompound nbt = new NBTCompound();
+		try {
+			serializable.serialize(nbt);
+		}
+		catch (Exception exception) {
+			exception.printStackTrace();
+			return null;
+		}
+		nbt.setString(NBT_CLASS_PATH_KEY, serializable.getClass().getName());
+        return nbt;
+    }
+
+    /**
+	 * Deserializes an {@link NBTCompound} into a registered {@link NBTSerializable} class, if it exists.
+	 *
+	 * @param <T> The class type that will be returned. It MUST match the class that was registered exactly.
+	 * @param nbt The NBT data that's used to populate the new instance of the class.
+	 * @return Returns a populated instance of a class, or null. It will return null if 1) the given NBTCompound is null
+	 * or empty, 2) the class identifier cannot be found, 3) the class cannot be matched, or 4) there was an error in
+	 * casting the instance into the given generic type.
+	 * */
+	@SuppressWarnings("unchecked")
+    public static <T extends NBTSerializable> T deserialize(NBTCompound nbt) {
+        if (nbt == null || nbt.isEmpty()) {
+            return null;
+        }
+        String path = nbt.getString(NBT_CLASS_PATH_KEY);
+        nbt.remove(NBT_CLASS_PATH_KEY);
+		if (Strings.isNullOrEmpty(path)) {
+			return null;
+		}
+		Class<? extends NBTSerializable> clazz = registeredClasses.get(path);
+		if (clazz == null) {
+			return null;
+		}
+		T instance;
+		try {
+			instance = (T) clazz.newInstance();
+			instance.deserialize(nbt);
+		}
+		catch (Exception exception) {
```

Comment:
Should do more explicit exception handling here and not just use the lazy "Exception"

Reply:
Yeeaaah, I was kind of conflicted about this. Would a printing the exception suffice? I don't want the serialize and deserialize to be more lenient, especially since it'll likely be used within streams a lot, I know it already is in ItemExchange

Reply:
Multi-try-Catch `(catch InvocationException|NoSuchMethodException|BlaBlaException e)` all the ones explicitly thrown by the reflection crap, log them and then throw a custom exception for the calling code. Anything going wrong here is pretty much guaranteed to be a bug in the calling code and should cause a RuntimeException.

Reply:
I've kept the lazy catch all here and [here](https://github.com/CivClassic/CivModCore/pull/17#discussion_r366096900), because if anything throws an exception, then something has gone horribly wrong, so instead of just logging, a contextualised exception is thrown that contains the original exception.